### PR TITLE
Merged common code base for SqlUtil.cs

### PR DESCRIPTION
**Pull request by arellegue**
Merged SqlUtil.cs from .netcore and .netfx.

### CODE REVIEW

**Review by dauinsight — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 1]
```diff
@@ -3,11 +3,32 @@
...
+using System.Runtime.Versioning;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+#if NET6_0_OR_GREATER
+using System.Transactions;
+#endif
+#if NETFRAMEWORK
+using SysTx = System.Transactions;
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 2]
```diff
@@ -3,11 +3,32 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Collections.Concurrent;
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 3]
```diff
@@ -3,11 +3,32 @@
...
+#endif
+#if NETFRAMEWORK
+using SysTx = System.Transactions;
+#endif
+using Microsoft.Data.Common;
 
-// This file will be used to merge code from the NetFx and NetCore SqlUtil.cs in the future.
 namespace Microsoft.Data.SqlClient
 {
+
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 4]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+            if (IntPtr.Zero != handle)
+            {
+                // SQLBU 359301: Currently, the server exports different names for x86 vs. AMD64 and IA64.  Supporting both names
+                //  for now gives the server time to unify names across platforms without breaking currently-working ones.
+                //  We can remove the obsolete name once the server is changed.
+                if (IntPtr.Zero != SafeNativeMethods.GetProcAddress(handle, "_______SQL______Process______Available@0"))
+                {
+                    _inProc = true;
+                }
+                else if (IntPtr.Zero != SafeNativeMethods.GetProcAddress(handle, "______SQL______Process______Available"))
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 5]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+    }
+
+    internal static class AsyncHelper
+    {
+#if NET6_0_OR_GREATER
+        internal static Task CreateContinuationTask(Task task, Action onSuccess, Action<Exception> onFailure = null)
+#endif
+#if NETFRAMEWORK
+        internal static Task CreateContinuationTask(Task task, Action onSuccess, SqlInternalConnectionTds connectionToDoom = null, Action<Exception> onFailure = null)
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 6]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+                                onSuccess();
+                            }
+                            catch (Exception e)
+                            {
+                                completion.SetException(e);
+                            }
+                        }
+                    }
+#endif
+#if NET6_0_OR_GREATER
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 7]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        internal static void ContinueTaskWithState(Task task,
+            TaskCompletionSource<object> completion,
+            object state,
+            Action<object> onSuccess,
+            Action<Exception, object> onFailure = null,
+            Action<object> onCancellation = null,
+#if NET6_0_OR_GREATER
+            Func<Exception, Exception> exceptionConverter = null
+#endif
+#if NETFRAMEWORK
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 8]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+
+        internal static void ContinueTask(Task task,
+            TaskCompletionSource<object> completion,
+            Action onSuccess,
+            Action<Exception> onFailure = null,
+            Action onCancellation = null,
+#if NET6_0_OR_GREATER
+            Func<Exception, Exception> exceptionConverter = null
+#endif
+#if NETFRAMEWORK
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 9]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+                        try
+                        {
+                            onSuccess(state2);
+                        }
+                        catch (Exception e)
+                        {
+                            completion.SetException(e);
+                        }
+#endif
+                    }
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 10]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+#if NETFRAMEWORK
+            if (isSystemOp)
+            {
+                return ADP.ArgumentNull(TdsEnums.TCE_PARAM_MASTERKEY_PATH, StringsHelper.GetString(Strings.TCE_NullCertificatePathSysErr, validLocations[0], validLocations[1], @"/"));
+            }
+            else
+            {
+                return ADP.ArgumentNull(TdsEnums.TCE_PARAM_MASTERKEY_PATH, StringsHelper.GetString(Strings.TCE_NullCertificatePath, validLocations[0], validLocations[1], @"/"));
+            }
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 11]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+#if NETFRAMEWORK
+            if (isSystemOp)
+            {
+                return ADP.ArgumentNull(TdsEnums.TCE_PARAM_MASTERKEY_PATH, StringsHelper.GetString(Strings.TCE_NullCngPathSysErr, @"/"));
+            }
+            else
+            {
+                return ADP.ArgumentNull(TdsEnums.TCE_PARAM_MASTERKEY_PATH, StringsHelper.GetString(Strings.TCE_NullCngPath, @"/"));
+            }
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 12]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+#if NETFRAMEWORK
+            if (isSystemOp)
+            {
+                return ADP.Argument(StringsHelper.GetString(Strings.TCE_InvalidCertificatePathSysErr, actualCertificatePath, validLocations[0], validLocations[1], @"/"), TdsEnums.TCE_PARAM_MASTERKEY_PATH);
+            }
+            else
+            {
+                return ADP.Argument(StringsHelper.GetString(Strings.TCE_InvalidCertificatePath, actualCertificatePath, validLocations[0], validLocations[1], @"/"), TdsEnums.TCE_PARAM_MASTERKEY_PATH);
+            }
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 13]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        /// </summary>
+        internal static void EscapeIdentifier(StringBuilder builder, string name)
+        {
+            Debug.Assert(builder != null, "builder cannot be null");
+#if NET6_0_OR_GREATER
+            Debug.Assert(!string.IsNullOrEmpty(name), "null or empty identifiers are not allowed");
+#endif
+#if NETFRAMEWORK
+            Debug.Assert(!ADP.IsEmpty(name), "null or empty identifiers are not allowed");
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 14]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        /// <param name="input">unescaped string</param>
+        /// <returns>escaped and quoted literal string</returns>
+        internal static string MakeStringLiteral(string input)
+        {
+#if NET6_0_OR_GREATER
+            if (string.IsNullOrEmpty(input))
+#endif
+#if NETFRAMEWORK
+            if (ADP.IsEmpty(input))
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 15]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+#if NETFRAMEWORK
+        private static readonly Lazy<MethodInfo> _enlistPromotableSinglePhase = new Lazy<MethodInfo>(() =>
+            typeof(SysTx.Transaction).GetMethod("EnlistPromotableSinglePhase", new Type[] { typeof(SysTx.IPromotableSinglePhaseNotification), typeof(Guid) }));
+
+        private static readonly Lazy<MethodInfo> _setDistributedTransactionIdentifier = new Lazy<MethodInfo>(() =>
+            typeof(SysTx.Transaction).GetMethod("SetDistributedTransactionIdentifier", new Type[] { typeof(SysTx.IPromotableSinglePhaseNotification), typeof(Guid) }));
+
+        private static readonly Lazy<MethodInfo> _getPromotedToken = new Lazy<MethodInfo>(() =>
+            typeof(SysTx.Transaction).GetMethod("GetPromotedToken"));
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 16]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        /// <param name="name">name of the identifier, in unescaped form</param>
+        /// <returns>escapes the name with [], also escapes the last close bracket with double-bracket</returns>
+        internal static string EscapeIdentifier(string name)
+        {
+#if NET6_0_OR_GREATER
+            Debug.Assert(!string.IsNullOrEmpty(name), "null or empty identifiers are not allowed");
+#endif
+#if NETFRAMEWORK
+            Debug.Assert(!ADP.IsEmpty(name), "null or empty identifiers are not allowed");
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 17]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+            else
+            {
+                return ADP.Argument(StringsHelper.GetString(Strings.TCE_InvalidCertificatePath, actualCertificatePath, validLocations[0], validLocations[1], @"/"), TdsEnums.TCE_PARAM_MASTERKEY_PATH);
+            }
+#endif
+        }
+
+        internal static Exception InvalidCspPath(string masterKeyPath, bool isSystemOp)
+        {
+#if NET6_0_OR_GREATER
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 18]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+#if NETFRAMEWORK
+            if (isSystemOp)
+            {
+                return ADP.Argument(StringsHelper.GetString(Strings.TCE_LargeCertificatePathLengthSysErr, actualLength, maxLength), TdsEnums.TCE_PARAM_MASTERKEY_PATH);
+            }
+            else
+            {
+                return ADP.Argument(StringsHelper.GetString(Strings.TCE_LargeCertificatePathLength, actualLength, maxLength), TdsEnums.TCE_PARAM_MASTERKEY_PATH);
+            }
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 19]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+#if NETFRAMEWORK
+            if (isSystemOp)
+            {
+                return ADP.ArgumentNull(TdsEnums.TCE_PARAM_ENCRYPTION_ALGORITHM, StringsHelper.GetString(Strings.TCE_NullKeyEncryptionAlgorithmSysErr));
+            }
+            else
+            {
+                return ADP.ArgumentNull(TdsEnums.TCE_PARAM_ENCRYPTION_ALGORITHM, StringsHelper.GetString(Strings.TCE_NullKeyEncryptionAlgorithm));
+            }
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 20]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+#if NETFRAMEWORK
+            if (isSystemOp)
+            {
+                return ADP.Argument(StringsHelper.GetString(Strings.TCE_InvalidKeyEncryptionAlgorithmSysErr, encryptionAlgorithm, validEncryptionAlgorithm), TdsEnums.TCE_PARAM_ENCRYPTION_ALGORITHM);
+            }
+            else
+            {
+                return ADP.Argument(StringsHelper.GetString(Strings.TCE_InvalidKeyEncryptionAlgorithm, encryptionAlgorithm, validEncryptionAlgorithm), TdsEnums.TCE_PARAM_ENCRYPTION_ALGORITHM);
+            }
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 21]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        internal static Exception CR_NextAttemptWillExceedQueryTimeout(SqlException innerException, Guid connectionId)
+        {
+            SqlErrorCollection errors = new SqlErrorCollection();
+            errors.Add(new SqlError(0, 0, TdsEnums.MIN_ERROR_CLASS, null, StringsHelper.GetString(Strings.SQLCR_NextAttemptWillExceedQueryTimeout), "", 0));
+#if NET6_0_OR_GREATER
+            SqlException exc = SqlException.CreateException(errors, "", connectionId, innerException);
+#endif
+#if NETFRAMEWORK
+            SqlException exc = SqlException.CreateException(errors, "", connectionId, innerException, batchCommand: null);
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 22]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        }
+
+        internal static Exception UnsupportedSysTxForGlobalTransactions()
+        {
+#if NET6_0_OR_GREATER
+            return ADP.InvalidOperation(StringsHelper.GetString(Strings.SQL_UnsupportedSysTxVersion));
+#endif
+#if NETFRAMEWORK
+            return ADP.InvalidOperation(StringsHelper.GetString(Strings.GT_UnsupportedSysTxVersion));
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 23]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+            }
+#endif
+#if NETFRAMEWORK
+        static internal SysTx.TransactionPromotionException PromotionFailed(Exception inner)
+        {
+            SysTx.TransactionPromotionException e = new SysTx.TransactionPromotionException(StringsHelper.GetString(Strings.SqlDelegatedTransaction_PromotionFailed), inner);
+            ADP.TraceExceptionAsReturnValue(e);
+            return e;
+        }
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 24]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        internal static Exception ChangePasswordConflictsWithSSPI()
+        {
+            return ADP.Argument(StringsHelper.GetString(Strings.SQL_ChangePasswordConflictsWithSSPI));
+        }
+        internal static Exception ChangePasswordRequires2005()
+        {
+            return ADP.InvalidOperation(StringsHelper.GetString(Strings.SQL_ChangePasswordRequiresYukon));
+        }
+#if NETFRAMEWORK
+        internal static Exception UnknownSysTxIsolationLevel(SysTx.IsolationLevel isolationLevel)
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 25]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        }
+        internal static Exception InvalidSQLServerVersionUnknown()
+        {
+            return ADP.DataAdapter(StringsHelper.GetString(Strings.SQL_InvalidSQLServerVersionUnknown));
+        }
+        internal static Exception SynchronousCallMayNotPend()
+        {
+            return new Exception(StringsHelper.GetString(Strings.Sql_InternalError));
+        }
+#if NET6_0_OR_GREATER
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 26]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        }
+        internal static Exception InvalidSSPIPacketSize()
+        {
+            return ADP.Argument(StringsHelper.GetString(Strings.SQL_InvalidSSPIPacketSize));
+        }
+        static internal Exception NullEmptyTransactionName()
+        {
+            return ADP.Argument(StringsHelper.GetString(Strings.SQL_NullEmptyTransactionName));
+        }
+#if NETFRAMEWORK
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 27]
```diff
@@ -27,6 +48,2993 @@ internal static ArgumentNullException ArgumentNull(string paramName)
...
+        }
+
+        internal static Task CreateContinuationTask<T1, T2>(Task task, Action<T1, T2> onSuccess, T1 arg1, T2 arg2, SqlInternalConnectionTds connectionToDoom = null, Action<Exception> onFailure = null)
+        {
+#if NET6_0_OR_GREATER
+            return CreateContinuationTask(task, () => onSuccess(arg1, arg2), onFailure);
+#endif
+#if NETFRAMEWORK
+            return CreateContinuationTask(task, () => onSuccess(arg1, arg2), connectionToDoom, onFailure);
+#endif
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlUtil.cs]: Applied suggestion.

**Review by DavoudEshtehari — APPROVED** [review]

**Pull request merged**